[WIP] Model trees (M5P)

smarie opened the pull request:
This PR adds M5P model (model trees) to scikit-learn, as mentioned in #13106.

It is still work in progress. The two example files in the `examples/` folder "work" already so you can try them, but there are **many** things that need to be thoroughly reviewed/tested/improved.

#### Implementation philosophy

It is an attempt to leverage the Cythonized tree growing procedure from scikit-learn, while implementing the M5P algorithm from Quilan (M5) +Wang (M5P), inspired by the M5P Weka classes. Note that there are significant discrepanties between the Weka implementation choices and the papers. But since Weka is definitely the reference for this kind of models, I am wondering if we should maybe offer both sets of parameters to let the users choose.

Note that other reference implementations also exist in R (cubist) and in python (LearningX).

#### Scientific references

Ross J. Quinlan: Learning with Continuous Classes. In: 5th Australian Joint Conference on Artificial Intelligence, Singapore, 343-348, 1992.

Y. Wang, I. H. Witten: Induction of model trees for predicting continuous classes. In: Poster papers of the 9th European Conference on Machine Learning, 1997.

#### Reference Issues
Fixes #13106

Commit:
Initial commit for the work on M5 models. The two example files work already, but there are many things that need to be thoroughly reviewed/tested.

edelserg commented:
I am just trying to find m5p on a python. I used to use it in Java. It would be good to have this work completed. And yet, the lack of the original algorithm is the inability to isolate variables that are included in the linear model of nodes, it is built for all possible variables and it is impossible to use categorical variable. If it is possible to do this would be good. I have done this for one of my old my project on c# there is no principle problems on this way...Good luck!

smarie commented:
Thanks for the encouragement ! I plan to come back to this later this summer. If I remember well, categorical variables were accepted in the Weka M5P in java. We used it a lot from within rapidminer, it was quite efficient and interpretable.

smarie commented:
Note for the future: I have improved my version of this in our industrial product, it now covers the "smoothing" feature described in the paper. That is not an excuse, I still need to find time to do a major refactoring of this to make it sklearn-y :) .

NitroFrench commented:
Hi guys, no news on this PR? This would be a tremendous addition to Scikit IMHO.

jnothman commented:
I agree that this could be a very exciting feature. As @smarie notes there's a lot of basic cleanup work here to make it fit in. @smarie, in what way could you accept help here?

smarie commented:
Hi @jnothman, this is in my backlog and to be honest, I was not willing to put so much energy on it until I have clear signs that it can actually move forward. From experience in other PRs (the kPCA one in particular), this can take some time. 

I am going to leave for paternity in a few days now, so my best target would be december. It can go extremely fast, as long as we agree on a few design guidelines. 

The main question from far is "should the pruning procedure be Cythonized for such a PR be accepted ?". If the answer is "not for the first version" then december is a realistic target. Note that this procedure is currrently here: https://github.com/scikit-learn/scikit-learn/pull/13732/files#diff-e7b2d516553f9d8cf4b14508a5c9c16aR714

RiverShah commented:
Hi, any updates on this please? Will be a much needed feature. Thanks!

hanshupe commented:
> Hi, any updates on this please? Will be a much needed feature. Thanks!

I only can confirm that, it's an urgently needed feature for me. Hope it can be included soon. Thanks!

smarie commented:
Note that you can try the code in this PR, it should work "as is" already

hanshupe commented:
> LearningX

One question: What is LearningX? I couldn't find anything about it... Thanks

smarie commented:
@hanshupe I edited my comment above. To answer your question, see link for LearningX at the beginning of thread #13106

hanshupe commented:
Any news on that topic, when Model trees will be available in sklearn?

hanshupe commented:
I would like to test the current M5 implementation. Could you please explain how I can get the correct sklearn version to test it? Is there some plan when M5 will be available?

Commit:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into M5P

 Conflicts:
	sklearn/tree/__init__.py

Commit:
Updated code from internal code base. Several bug fixes + smoothing capability. Also new capability to turn off pruning with `use_pruning=False`

Commit:
Added M5P tests and fixed a bug with numpy broadcasting/ravel

smarie commented:
@hanshupe Great to hear that your interest is still alive !

I guess that the easiest way for you to perform this test is to

 - clone my fork and checkout this "M5P" branch https://github.com/smarie/scikit-learn/tree/M5P
 - rebuild sklearn from this source folder.

I just updated the branch and added the smoothing capability, as well as the "no pruning" flag. The included example `plot_tree_regression_m5.py` should work and display this figure:

![image](https://user-images.githubusercontent.com/3236794/111353014-22f22900-8685-11eb-980a-4104f113d022.png)

As well as display the equations, for example for the last tree :

```
----- Tree 3 (smoothing_constant_ratio=0.001, use_pruning=True)
M5Prime (pre-smoothed with constant 15):
X[0] <= 3.133 [friedman_mse=0.467, samples=80] (err=0.184, params=17)
|   X[0] <= 0.514 [friedman_mse=0.09, samples=51] (err=0.215, params=11)
|   |   LEAF [friedman_mse=0.035, samples=11] : LM1 (err=0.317, params=2)
|   |   X[0] <= 2.516 [friedman_mse=0.05, samples=40] (err=0.177, params=8)
|   |   |   X[0] <= 0.89 [friedman_mse=0.02, samples=32] (err=0.191, params=5)
|   |   |   |   LEAF [friedman_mse=0.006, samples=9] : LM2 (err=0.09, params=2)
|   |   |   |   LEAF [friedman_mse=0.011, samples=23] : LM3 (err=0.218, params=2)
|   |   |   LEAF [friedman_mse=0.01, samples=8] : LM4 (err=0.106, params=2)
|   X[0] <= 3.85 [friedman_mse=0.084, samples=29] (err=0.107, params=5)
|   |   LEAF [friedman_mse=0.018, samples=14] : LM5 (err=0.076, params=2)
|   |   LEAF [friedman_mse=0.007, samples=15] : LM6 (err=0.13, params=2)
LM1: 2.931e-01 * X[0] + 4.314e-01
LM2: 4.829e-02 * X[0] + 6.798e-01
LM3: -9.365e-02 * X[0] + 8.489e-01
LM4: -2.314e-01 * X[0] + 1.091
LM5: -5.663e-01 * X[0] + 1.645
LM6: -3.755e-01 * X[0] + 8.723e-01
```

The second example in `plot_tree_regression_m5_boston_housing.py` shows how the M5 tree (bottom row) compares to the gradient boosting regressor (top row) on the boston housing dataset, and shows how the tree can be compressed to only work on datasets containing the features that it actually uses. It also displays some feature importance but I'm not sure the heuristic I came up with for this computation is very good :) (none is mentioned in the original papers)

![image](https://user-images.githubusercontent.com/3236794/111354047-30f47980-8686-11eb-8133-8a1fd6b2674b.png)

thomasjpfan commented:
> do you know if it would be mandatory to cythonize the pruning procedure as asked above ? Who in the maintainers' team should be my contact to review subsequent changes to this PR ?

I would be someone you can ping for Cython reviews. Before we go down this route, how long does it take to fit when compared to `DecisionTreeRegressor`? It would interesting to see how the current implementation performs with/without smoothing and with/without pruning. This would help us determine if we should write this in Cython.

smarie commented:
I realize from #13106 that my "no pruning" implementation differs from the one a user would expect, as it defaults to the underlying decision tree without using the linear model on the leaves. I should correct this.

cab938 commented:
@smarie given the speed at which this would make it back into sklearn, do you happen to have an impl to share which is stand alone?

cerlymarco commented:
@cab938 I can suggest you also [linear-tree](https://github.com/cerlymarco/linear-tree): a python library to build Model Trees with Linear Models at the leaves.

Iicolan commented:
@smarie hi, I would like to find the best value of "smoothing_constant_ratio" via GridSearchCV, but it always returned "FitFailedWarning" as follows, could you please have a look of this problem?

scikit-learn/sklearn/model_selection/_validation.py:617: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "scikit-learn/sklearn/model_selection/_validation.py", line 596, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/scikit-learn/sklearn/tree/_classes_m5.py", line 246, in fit
    self.node_models, global_std_dev, use_pruning=self.use_pruning
  File "/scikit-learn/sklearn/tree/_classes_m5.py", line 1250, in build_models_and_get_pruning_info
    y_true_this_node = y_train_all[samples_at_this_node]

smarie commented:
Hi @cab938 sorry for the delay in responding ! This is a good idea, maybe I'll do both. Unfortunately however I'll have very few bandwidth to work on this until summer. So the fastest route for you is probably to copy the classes from the pull request locally, as they currently do not contain any cythonization for now - they are standalone.

Sorry for the inconvenience - this is done on "free time" and sometimes free time is hard to find :)

cab938 commented:
> Hi @cab938 sorry for the delay in responding ! This is a good idea, maybe I'll do both. Unfortunately however I'll have very few bandwidth to work on this until summer. So the fastest route for you is probably to copy the classes from the pull request locally, as they currently do not contain any cythonization for now - they are standalone.
> 
> Sorry for the inconvenience - this is done on "free time" and sometimes free time is hard to find :)

NP, I did do as suggested and just ripped things out and it seems to work well, so now I'm going to build a lecture on these methods!

MarjoleindeWith commented:
Hi @smarie,
I love your work! I have a question: do you know of any way to make the code run faster? I have a data set with 10.000 observations and ca. 60 variables, but it takes incredibly long to build just one tree. Hope you can help me!

Kind regards,
Marjolein

hanshupe commented:
> Hi @smarie,
> I love your work! I have a question: do you know of any way to make the code run faster? I have a data set with 10.000 observations and ca. 60 variables, but it takes incredibly long to build just one tree. Hope you can help me!
> 
> Kind regards,
> Marjolein

I have the same problem, so i only use max_depth = 2. Besides that I like the work, hope to see it in sklearn soon!

smarie commented:
@MarjoleindeWith and @hanshupe  thanks to both of you for your feedback ! Well I guess that this answers the question from @thomasjpfan  about cythonizing the code :) : although the implementation is certainly suboptimal at the moment, I do not expect a huge gain if we stay in pure python. 

Another thing that you can do to accelerate the fit time is setting `use_smoothing` to `False` or to `'on_prediction'`. Note that however the latter will make predictions much slower.

I'll do my best to find a few days to take another stab at this during summer, most probably mid august.

MarjoleindeWith commented:
> @MarjoleindeWith and @hanshupe thanks to both of you for your feedback ! Well I guess that this answers the question from @thomasjpfan about cythonizing the code :) : although the implementation is certainly suboptimal at the moment, I do not expect a huge gain if we stay in pure python.
> 
> Another thing that you can do to accelerate the fit time is setting `use_smoothing` to `False` or to `'on_prediction'`. Note that however the latter will make predictions much slower.
> 
> I'll do my best to find a few days to take another stab at this during summer, most probably mid august.

Thank you very much for responding! I have another short question I hope you don't mind answering: for the Logistic Model Trees, did you use the exact same approach of Landwehr et al.? That is, including LogitBoost, where the parent nodes are the basis for the logistic regression at the child's nodes? Or did you use the 'normal' approach from Quinlan, but then using Logistic Regression instead of Linear Regression? Thank you very much and have a good weekend!

Kind regards,
Marjolein

smarie commented:
> for the Logistic Model Trees, did you use the exact same approach of Landwehr et al.? That is, including LogitBoost, where the parent nodes are the basis for the logistic regression at the child's nodes? Or did you use the 'normal' approach from Quinlan, but then using Logistic Regression instead of Linear Regression?

The current Pull request does not support this kind of model: indeed, although you **can** set the leaf model to be any model in the contructor,
 
 - pruning is made based on perf metrics that is `root_mean_squared_error`, so only regression leaf models can possibly work. See [here](https://github.com/scikit-learn/scikit-learn/blob/8860facaeab4579dea25bb5b32b8e2376dc502c8/sklearn/tree/_classes_m5.py#L410)

 - and smoothing with "pre-installed" mode is definitely relying on linear regression coefficients ([here](https://github.com/scikit-learn/scikit-learn/blob/8860facaeab4579dea25bb5b32b8e2376dc502c8/sklearn/tree/_classes_m5.py#L425))

But maybe I misunderstood your question ?

Iicolan commented:
Hi, @smarie,
    very appreciate for your work! and recently I met a problem when I try to use GridSearchCV to find the best value of "smoothing_constant_ratio" . It returned "FitFailedWarning" no matter what score metrics I used. Could you please help me have a look at this one, or which score metric does M5P support in python code? Thank you very much!

Best Regard.

smarie commented:
Hi @licolan thanks for the feedback ! I do not have any idea about the origin of such warning. This may be something to investigate with a minimum reproducible dataset/example. Concerning the metric, any regression metric should work so rmse, r², etc. should be ok.

Iicolan commented:
> Hi @licolan thanks for the feedback ! I do not have any idea about the origin of such warning. This may be something to investigate with a minimum reproducible dataset/example. Concerning the metric, any regression metric should work so rmse, r², etc. should be ok.

Thank you very much for your reply!

Iicolan commented:
> Hi @licolan thanks for the feedback ! I do not have any idea about the origin of such warning. This may be something to investigate with a minimum reproducible dataset/example. Concerning the metric, any regression metric should work so rmse, r², etc. should be ok.

Thanks @smarie for your reply! The problem has been solved, and thanks for your nice reply again!

smarie commented:
Hi there, this morning I presented the new `m5py` package at PyConDE PyData. A first release has been made, and a few examples are available here: https://smarie.github.io/python-m5p/ 
Note that the example figures are available as thumbnails but not in the notebooks, due to a temporary bug in mkdocs-gallery.

Do not hesitate to try and contribute ! Thanks for your interest in this topic. I am convinced that this is something useful for the community even if the algorithm itself is not very new.

MarianodelRio commented:
Hi guys, are there any news about this PR?

smarie commented:
@MarianodelRio : this PR is now closed, the official version is now in https://smarie.github.io/python-m5p/

smarie commented:
@MarianodelRio I suggest that we discuss this here : https://github.com/scikit-learn/scikit-learn/issues/13106
